Introduce type sugar for InlineArray

hamishknight opened the pull request:
Parse e.g `[3 x Int]` as type sugar for InlineArray. Gated behind an experimental feature flag for now.

hamishknight commented:
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test

xedin commented on lib/Sema/CSSolver.cpp:
```diff
@@ -1201,14 +1209,15 @@ void ConstraintSystem::shrink(Expr *expr) {
 
       case TypeReprKind::Array:
       case TypeReprKind::Dictionary:
+      case TypeReprKind::InlineArray:
```
At this point I think we shouldn't expand shrink to new types, let's let the proper solver handle it...

hamishknight replied:
Fair enough, should we also decline to handle the non-sugared version?

hamishknight commented:
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test

adrian-prantl commented on lib/IRGen/IRGenDebugInfo.cpp:
```diff
@@ -2253,6 +2253,7 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     // SyntaxSugarType derivations.
     case TypeKind::Dictionary:
     case TypeKind::ArraySlice:
+    case TypeKind::InlineArray:
```
You will need to update SwiftASTContext.cpp in LLDB, too.

adrian-prantl commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1509,6 +1509,14 @@ void ASTMangler::appendType(Type type, GenericSignature sig,
+    case TypeKind::InlineArray: {
+      assert(DWARFMangling && "sugared types are only legal for the debugger");
+      // FIXME: Once we have a known mangling substitution for InlineArray,
+      // use that. For now just desugar.
```
We'll need a mangling for the sugar for the ForDebugger mangling kind.

adrian-prantl replied:
> Is there any harm in mangling it as the desugared type for now?

1. nobody will come and fix this FIXME later, and you have all the necessary context in memory right now
2. A type that is spelled as `[3 x Int]` in source code will show up as an `InlineArray<3, Int>` in the debugger, which is confusing for anyone not familiar with the implementation
3. LLDB data formatters match on type names, so this has an effect on how people need to register their custom formatters. This means it would be higher friction to change this later.

adrian-prantl replied:
You can probably just copy&paste the Array sugar mangling (with a different letter)

hamishknight replied:
Right, I was hoping we could use the same known-type mangling as InlineArray though (to match the other sugared manglings), which doesn't currently exist. That being said, looking at the known-type manglings, it seems like we've actually exhausted all the available single-letter manglings and would need to add a 3rd level of known-type manglings if we wanted to add InlineArray. So maybe we should just pick an arbitrary letter for the sugared mangling for now (`A` seems fine, seems unlikely `AutoreleasingUnsafeMutablePointer` will ever get a sugar type)

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/10291
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test

hamishknight committed:
[Sema] Sink `ValueMatchVisitor` into `applyUnboundGenericArguments`

Make sure it's called for sugar code paths too. Also let's just always
run it since it should be a pretty cheap check.

hamishknight committed:
[CS] Decline to handle InlineArray in shrink

Previously we would try the contextual type `(<int>, <element>)`,
which is wrong. Given we want to eliminate shrink, let's just bail.

hamishknight committed:
[Parse] Split up `canParseType` 

While here, address the FIXME in `canParseTypeSimpleOrComposition`
and only check to see if we can parse a type-simple, including
`each`, `some`, and `any` for better recovery.

hamishknight committed:
[Sema] Diagnose passing integer to non-integer type parameter

This was previously missed, though would have been diagnosed later
as a requirement failure.

hamishknight committed:
Introduce type sugar for InlineArray

Parse e.g `[3 x Int]` as type sugar for InlineArray. Gated behind
an experimental feature flag for now.

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/10291
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test

hamishknight commented:
https://github.com/swiftlang/llvm-project/pull/10291
https://github.com/swiftlang/swift-syntax/pull/3021

@swift-ci please test
